fix: Improve Query Consumer Event Handling Performance

igdianov opened the pull request:
- [x] Optimized entities hash code and equals to be consistent using entity identifier 
- [x] Changed Audit entity identity generator to use sequence generator
- [x] Changed Query variable entity identity generator to use sequence generator
- [x] Enabled Hibernate query batching for Audit and Query event handlers
- [x] Optimized Audit event entities to be immutable to avoid extra update queries 
- [x] Optimized Audit event entities to use dynamic insert query behavior
- [x] Optimized Query event entities to use dynamic insert and update query behavior 
- [x] Replace Spring Data Repository overhead with JPA EntityManager 
- [x] Optimized Query event handler entity graph fetching to reduce N+1 select queries
- [x] Optimized task variables and task created events handling order

igdianov committed:
fix(audit): update pk generated value strategy to use sequence

igdianov committed:
fix(audit): optimize equals and hash code in persistent entities

igdianov committed:
fix(audit): add batch executor for audit entity handling

igdianov committed:
fix(query): optimize query entities equals and hash code

igdianov committed:
fix(notifications): update data.sql

igdianov committed:
fix(audit): use consistent equals and hash code

igdianov committed:
fix(audit): update unit tests

igdianov committed:
fix(audit): refactor AuditConsumerChannelHandlerImpl class

igdianov committed:
fix(query): use sequence generator for variable instance entity

igdianov committed:
fix(query): use consistent equals and hash code for query entities

igdianov committed:
fix(query): initial query event handler batching optimization

igdianov committed:
fix(query): refactor task event handlers to use entity manager

igdianov committed:
fix(query): refactor process entity event handlers to use entity manager

igdianov committed:
fix(query): optimize application deployed event handler

igdianov committed:
fix(query): replace get reference with find method for process started

igdianov committed:
fix(query): imporove BaseBPMNActivityEntity equals

igdianov committed:
fix(query): update bpmn activity event handlers

igdianov committed:
fix(query): optimize integration context event handlers

igdianov committed:
fix(query): optimize bpmn sequence taken event handler

igdianov committed:
fix(query): add dirties context to QuerySwaggerITSupport

igdianov committed:
fix(audit): refine batch executor strategy

igdianov committed:
fix(query): add test datasource proxy logger

igdianov committed:
fix(query): add test jpa hibernate batch config

igdianov committed:
fix(query): update jpa entity models

igdianov committed:
fix(query): fetch entity graph in query consumer message handler

igdianov committed:
fix(query): refactor variable event handlers to use fetched entity graph

igdianov committed:
fix(query): refactor bpmn activity event handlers to use fetched entity graph

igdianov committed:
fix(query): refactor integration context event handlers to use fetch graph

igdianov committed:
fix(query): remove query event handler context transaction requirement

igdianov committed:
fix(query): update event handler unit tests

igdianov committed:
fix(query): optimize event handler for process deployed events

igdianov committed:
fix(query): optimize task candidate user / group event handlers

igdianov committed:
fix(query): extract QueryEntityGraphFetchingOptimizer in separate class

igdianov committed:
fix(query): revert QueryEventHandlerContext changes

igdianov committed:
fix(query): better name for query event handler context optimizer

igdianov committed:
fix(audit): declare explicit sequence generator

igdianov committed:
fix(audit): update variable value hash code and equals

igdianov committed:
fix(audit): replace batch executor with Spring repository saveAll

igdianov committed:
fix(audit): make audit event entity immutable to avoid dirty checks

igdianov committed:
fix(audit): add datasource-proxy-spring-boot-starter for tests

igdianov committed:
fix(query): add hibernate.generate_statistics=true in test properties

igdianov committed:
fix: optimize queries using Hibernate dynamic insert and update annotations

erdemedeiros commented on activiti-cloud-audit-service/activiti-cloud-services-audit-jpa/src/main/java/org/activiti/cloud/services/audit/jpa/conf/AuditJPAAutoConfiguration.java:
```diff
@@ -15,57 +15,16 @@
-import org.activiti.cloud.services.audit.jpa.converters.VariableCreatedEventConverter;
-import org.activiti.cloud.services.audit.jpa.converters.VariableDeletedEventConverter;
-import org.activiti.cloud.services.audit.jpa.converters.VariableUpdatedEventConverter;
+import org.activiti.cloud.services.audit.jpa.converters.*;
```
Please, check your imports configuration. We don't use `.*` in imports.

igdianov replied:
fixed

erdemedeiros commented on activiti-cloud-audit-service/activiti-cloud-services-audit-jpa/src/main/java/org/activiti/cloud/services/audit/jpa/events/ActivityCancelledAuditEventEntity.java:
```diff
@@ -16,15 +16,14 @@
 
 @Entity(name = ActivityCancelledAuditEventEntity.ACTIVITY_CANCELLED_EVENT)
 @DiscriminatorValue(value = ActivityCancelledAuditEventEntity.ACTIVITY_CANCELLED_EVENT)
+@DynamicInsert
```
@igdianov, have you noticed important performance improvements with the addition of this annotation? There are some [controversy about using it or not](https://stackoverflow.com/questions/21721818/why-does-not-hibernate-set-dynamicinsert-by-default), including a comment saying that it breaks batch updates.

igdianov replied:
The performance improvements are noticeable for audit events due to their specific attribute mappings. I will test performance for query event with and without dynamic sql statements to be sure

igdianov replied:
The dynamic insert performs better than default behavior

erdemedeiros commented on activiti-cloud-audit-service/activiti-cloud-services-audit-jpa/src/main/java/org/activiti/cloud/services/audit/jpa/events/AuditEventEntity.java:
```diff
@@ -225,24 +222,7 @@ public void setMessageId(String messageId) {
-                            serviceType,
-                            serviceVersion,
-                            timestamp);
+        return getClass().hashCode();
```
The `hashCode` method should be consistent with the `equals` method. So you should be using the `id` to calculate the hash as you are using it to calculate equatility.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-audit-service/activiti-cloud-services-audit-jpa/src/main/java/org/activiti/cloud/services/audit/jpa/events/VariableValue.java:
```diff
@@ -41,15 +41,19 @@ public boolean equals(Object o) {
     }
 
     @Override
     public int hashCode() {
```
I think the previous implementation is the right one as it should be consistent with the `equals` method.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-audit-service/activiti-cloud-services-audit-jpa/src/main/java/org/activiti/cloud/services/audit/jpa/streams/AuditConsumerChannelHandlerImpl.java:
```diff
@@ -53,16 +58,22 @@ public AuditConsumerChannelHandlerImpl(EventsRepository eventsRepository,
+                                                                                .name());
                 }
             }
+            // TODO: test and implement flush for batch size
```
Is this `TODO` still needed?

igdianov replied:
fixed

erdemedeiros commented on activiti-cloud-audit-service/activiti-cloud-starter-audit/src/test/resources/application.properties:
```diff
@@ -16,3 +16,27 @@ keycloak.security-constraints[0].authRoles[0]=ACTIVITI_USER
 keycloak.security-constraints[1].authRoles[0]=ACTIVITI_ADMIN
 keycloak.security-constraints[1].securityCollections[0].patterns[0]=/admin/*
+
+spring.jpa.properties.hibernate.jdbc.batch_size=50
```
Should this be set for the main application as well (in addition to tests)?

igdianov replied:
fixed

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-liquibase/src/main/resources/config/query/liquibase/changelog/13-alter.pg.schema.7.1.0-m17.sql:
```diff
@@ -0,0 +1,3 @@
+create sequence variable_sequence start with 1 increment by 50;
+
+select setval('variable_sequence', (select max(id) from process_variable));
```
Shouldn't this also take into account `task_variable` table? What about the column type? it should be changed from `bigserial`, right?

igdianov replied:
Good catch! The sequence generator should be separate for process and task variables.

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-liquibase/src/main/resources/config/query/liquibase/changelog/13-alter.pg.schema.7.1.0-m17.sql:
```diff
@@ -0,0 +1,3 @@
+create sequence variable_sequence start with 1 increment by 50;
```
Missing alter table script for oracle

igdianov replied:
Fixed

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/AbstractVariableEntity.java:
```diff
@@ -15,29 +15,19 @@
 
     @Id
-    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @GeneratedValue(generator = "variable_sequence", strategy = GenerationType.SEQUENCE)
```
Shouldn't it include `@SequenceGenerator` as you did for audit?

igdianov replied:
fixed

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/AbstractVariableEntity.java:
```diff
@@ -177,22 +167,22 @@ public void setProcessInstance(ProcessInstanceEntity processInstance) {
-        int result = super.hashCode();
-        result = prime * result + Objects.hash(id);
-        return result;
+        return getClass().hashCode();
```
It should use `id` to calculate the hash.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/ActivitiEntityMetadata.java:
```diff
@@ -98,23 +97,26 @@ public void setServiceType(String serviceType) {
     @Override
     public int hashCode() {
-        return Objects.hash(appName, appVersion, serviceFullName, serviceName, serviceType, serviceVersion);
+        return getClass().hashCode();
```
It should be consistent with the `equals` method.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/ApplicationEntity.java:
```diff
@@ -63,26 +68,24 @@ public String getName() {
-        int result = super.hashCode();
-        result = prime * result + Objects.hash(id);
-        return result;
+        return getClass().hashCode();
```
It should be consistent with `equals`.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/BPMNSequenceFlowEntity.java:
```diff
@@ -154,163 +151,98 @@ public String getProcessInstanceId() {
-                                               targetActivityName,
-                                               targetActivityType);
-        return result;
+        return getClass().hashCode();
```
It should be consistent with `equals`.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/BaseBPMNActivityEntity.java:
```diff
@@ -201,89 +200,23 @@ public void setExecutionId(String executionId) {
-                                               startedDate,
-                                               status);
-        return result;
+        return getClass().hashCode();
```
It should be consistent with `equals`.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/IntegrationContextEntity.java:
```diff
@@ -327,136 +314,26 @@ public void setErrorCode(String errorCode) {
-                                               stackTraceElements,
-                                               status);
-        return result;
+        return getClass().hashCode();
```
It should be be consistent with `equals`.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/ProcessDefinitionEntity.java:
```diff
@@ -113,10 +112,7 @@ public void setFormKey(String formKey) {
-        int result = super.hashCode();
-        result = prime * result + Objects.hash(id);
-        return result;
+        return getClass().hashCode();
```
It should be consistent with `equals`.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/ProcessInstanceEntity.java:
```diff
@@ -400,67 +403,23 @@ public void setSequenceFlows(List<BPMNSequenceFlowEntity> sequenceFlows) {
-                                               suspendedTo,
-                                               status);
-        return result;
+        return getClass().hashCode();
```
It should be consistent with `equals`.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/ProcessVariableEntity.java:
```diff
@@ -57,33 +62,38 @@ public ProcessVariableEntity(Long id,
     @Override
     public int hashCode() {
-        return super.hashCode();
+        return getClass().hashCode();
```
It should be consistent with `equals`.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/ServiceTaskEntity.java:
```diff
@@ -44,28 +59,22 @@ public void setIntegrationContext(IntegrationContextEntity integrationContext) {
     @Override
     public int hashCode() {
-        return super.hashCode();
+        return getClass().hashCode();
```
It should be consistent with `equals`.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/TaskCandidateGroupEntity.java:
```diff
@@ -85,18 +79,23 @@ public void setTask(TaskEntity task) {
     @Override
     public int hashCode() {
-        return Objects.hash(groupId, taskId);
+        return getClass().hashCode();
```
It should be consistent with `equals`.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/TaskCandidateGroupId.java:
```diff
@@ -41,12 +41,7 @@ public String getGroupId() {
-                + ((taskId == null) ? 0 : taskId.hashCode());
-        result = prime * result + groupId.hashCode();
-        return result;
+        return getClass().hashCode();
```
It should be consistent with `equals`. The previous implementation was fine.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/TaskCandidateUserEntity.java:
```diff
@@ -88,7 +92,7 @@ public void setTask(TaskEntity task) {
     @Override
     public int hashCode() {
-        return Objects.hash(taskId, userId);
+        return getClass().hashCode();
```
It should be consistent with `equals`. The previous implementation was fine.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/TaskCandidateUserId.java:
```diff
@@ -42,12 +42,7 @@ public String getUserId() {
-                + ((taskId == null) ? 0 : taskId.hashCode());
-        result = prime * result + userId.hashCode();
-        return result;
+        return getClass().hashCode();
```
It should be consistent with `equals`.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/TaskEntity.java:
```diff
@@ -557,10 +544,7 @@ public boolean isInFinalState(){
-        int result = super.hashCode();
-        result = prime * result + Objects.hash(id);
-        return result;
+        return getClass().hashCode();
```
It should be consistent with `equals`.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/TaskVariableEntity.java:
```diff
@@ -102,10 +99,7 @@ public boolean isTaskVariable() {
-        int result = super.hashCode();
-        result = prime * result + Objects.hash(taskId);
-        return result;
+        return getClass().hashCode();
```
It should be consistent with `equals`.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-model/src/main/java/org/activiti/cloud/services/query/model/VariableValue.java:
```diff
@@ -29,22 +29,19 @@ public VariableValue(T value) {
-		int result = 1;
-		result = prime * result + ((value == null) ? 0 : value.hashCode());
-		return result;
+		return getClass().hashCode();
```
It should be consistent with `equals`.

igdianov replied:
See comment about equals and hashCode for JPA entities: https://github.com/Activiti/activiti-cloud/pull/522#issuecomment-1001805745

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/BaseBPMNActivityEventHandler.java:
```diff
@@ -15,56 +15,74 @@
+
+    }
+
+    public <R> BaseBPMNActivityEntity createBpmnActivityEntity(CloudRuntimeEvent<?, ?> event) {
```
The generic parameter <R> is never used and can be removed.

igdianov replied:
fixed

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/BaseBPMNActivityEventHandler.java:
```diff
@@ -15,56 +15,74 @@
+
+        String pkId = BPMNActivityEntity.IdBuilder.from(bpmnActivity);
+
+        BaseBPMNActivityEntity bpmnActivityEntity = entityManager.find(BPMNActivityEntity.class,
```
This can generate duplicates in case where we have in-fly instances where `ActivityStarted` has already been processed (and created using the event id as id), but not `ActivitiCompleted`. In this case a new entry will be created in the database and the previous one will be left in `Started` status. I'm not sure about how to handle it.

igdianov replied:
It will leave orphaned entities for in-flight processes. There is no better to deal with this.

erdemedeiros replied:
I agree. We can always handle the specific cases later on, when/if they happen.

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/BaseIntegrationEventHandler.java:
```diff
@@ -31,58 +28,17 @@
-                logger.error("Cannot find BPMNActivityEntity for integrationContext: {}", integrationContext);
-            }
-        }
+        IntegrationContextEntity entity = entityManager.find(IntegrationContextEntity.class, pkId);
```
Same potential issue here with in-fly instances.

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-starter-query/pom.xml:
```diff
@@ -115,6 +115,12 @@
+    <dependency>
+      <groupId>com.github.gavlyukovskiy</groupId>
+      <artifactId>datasource-proxy-spring-boot-starter</artifactId>
+      <version>1.7.1</version>
```
It's probably better to manage the version inside a `dependencyManagement` section, so we only define it in single place for both audit and query.

igdianov replied:
@erdemedeiros I added it for testing purposes only. I removed it to avoid as unnecessary dependency.

erdemedeiros reviewed (changes requested):
Great job, @igdianov ! Find bellow some questions and suggestions.

igdianov committed:
fix: compile error in ProcessDeployedEventHandler

igdianov committed:
fix: add explicit `@SequenceGenerator` for AbstractVariableEntity id attribute

igdianov committed:
fix: remove TODO comment

igdianov committed:
fix: add Jdbc batching configuration to starter metadata properties

igdianov committed:
fix: use separate sequence generator for task and process variables

igdianov committed:
fix: add variable sequence declarations for Oracle DDL

igdianov commented:
@erdemedeiros I used the following article as a guide to implement hashCode and equals for persistent entities using entity identifer: https://vladmihalcea.com/how-to-implement-equals-and-hashcode-using-the-jpa-entity-identifier/


> When the entity identifier is null, we can guarantee equality only for the same object references. Otherwise, no transient object is equal to any other transient or persisted object. That’s why the identifier equality check is done only if the current Object identifier is not null.
> 
> With this implementation, the equals and hashCode work fine for all entity state transitions. The reason why it works is that the hashCode value does not change, hence, we can rely on the java.lang.Object reference equality as long as the identifier is null.

Equals and hashCode must behave consistently across all entity state transitions: https://vladmihalcea.com/a-beginners-guide-to-jpa-hibernate-entity-state-transitions/

igdianov committed:
fix: remove datasource decorator dependency

igdianov committed:
fix: use explicit import without wildcard

igdianov committed:
fix: The generic parameter is never used and can be removed

igdianov committed:
fix: polish code style and remove unused code

erdemedeiros commented:
@igdianov, thank you for the pointers for `equals` to `hashcode` methods. I understand better now!

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-liquibase/src/main/resources/config/query/liquibase/changelog/10-alter.oracle.schema.7.1.0-m17.sql:
```diff
@@ -1,2 +1,14 @@
+
+declare max_process_variable number;
+begin
+    select nvl(max(value), 0) + 1 into max_process_variable from process_variable;
```
Shouldn't it be `max(id)` instead of `max(value)`?

igdianov replied:
fixed

igdianov committed:
fix: update Oracle DDL for sequence variables

igdianov committed:
fix: update Oracle DDL syntax

igdianov committed:
fix: changed exception log level to error for VariableCreatedEventHandler

igdianov committed:
fix: extracted findVariableByName into TaskEntity

igdianov committed:
fix: handle task variable created event before task entity create event

igdianov committed:
fix: handle process variable created event before process instance created

igdianov committed:
fix: update VariableEntityCreatedEventHandlerTest

igdianov committed:
fix: update task and process variable handlers for acceptance tests

igdianov committed:
fix: update QueryTaskEntityVariablesIT to use correct events order

igdianov committed:
fix: remove code duplication in variable handlers

igdianov committed:
fix: add hibernate sql logging for tests

igdianov committed:
fix: optimize task variable created handler with entity graph

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-starter-query/src/test/resources/application.properties:
```diff
@@ -33,3 +33,5 @@ keycloak.security-constraints[1].authRoles[0]=ACTIVITI_ADMIN
 keycloak.security-constraints[1].securityCollections[0].patterns[0]=/admin/*
 
 spring.jpa.properties.hibernate.generate_statistics=true
+spring.jpa.properties.hibernate.format_sql=true
+spring.jpa.show-sql=true
```
Was this a temporary change? I'm afraid this will increase a lot the size of the logs.

igdianov replied:
It does not increase logs as long as the logger level is OFF for tests

erdemedeiros replied:
Ok, so we need to set it to OFF for Hibernate in the github actions because it's currently enabled https://github.com/Activiti/activiti-cloud/runs/4672967808?check_suite_focus=true#step:7:10434.

igdianov committed:
fix: refine query event handler context optimizer entity graph load

igdianov committed:
fix: update QueryEventHandlerContext to add sorted event handler

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/EntityManagerFinder.java:
```diff
@@ -0,0 +1,72 @@
+                                                                                   String processInstanceId) {
+        EntityGraph<ProcessInstanceEntity> entityGraph = entityManager.createEntityGraph(ProcessInstanceEntity.class);
+
+        entityGraph.addAttributeNodes("variables");
```
A constant can be extracted here.

igdianov replied:
Fixed

igdianov committed:
fix: add setval for audit sequence

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/EntityManagerFinder.java:
```diff
@@ -0,0 +1,72 @@
+
+public class EntityManagerFinder {
+
+    public static Optional<TaskEntity> findTaskWithVariables(EntityManager entityManager,
```
I'd prefer to have these methods as non-static and inject instances of `EntityManagerFinder` where needed, so it can be easily mocked if necessary.

igdianov replied:
Fixed

erdemedeiros commented on activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/test/java/org/activiti/cloud/services/query/events/handlers/ProcessVariableEntityDeletedHandlerTest.java:
```diff
@@ -57,8 +64,10 @@ public void handleRemoveVariableFromProcessAnDeleteIt() {
 
-        given(entityManager.find(ProcessInstanceEntity.class,
-                                "procInstId")).willReturn(processInstanceEntity);
+        when(entityManager.createEntityGraph(ProcessInstanceEntity.class)).thenReturn(mock(EntityGraph.class));
+        given(entityManager.find(eq(ProcessInstanceEntity.class),
+                                 eq("procInstId"),
+                                 any(Map.class))).willReturn(processInstanceEntity);
```
This will be simplified if we use non static methods in `EntityManagerFinder`:
```
given(entityManagerFinder.findProcessInstanceWithVariables(entitiyManager, "procInstId")).willReturn(processInstanceEntity);
```

igdianov replied:
Fixed

igdianov committed:
fix: add events order optimization for QueryEventHandlerContextOptimizer


fix: update event handling sort order to put candidate events after task updates

igdianov committed:
fix: refactor EntityManagerFinder implementation

igdianov committed:
fix: add QueryEventHandlerContextOptimizerTest

igdianov committed:
fix: extract constants

erdemedeiros committed:
Simplify EntityManagerFinder

- inject EntityManager via constructor, so it can be removed from the parameter list in the methods

erdemedeiros reviewed (approved):
LGTM!

erdemedeiros commented:
Ref https://github.com/Activiti/Activiti/issues/3803.
